### ISSUE TITLE
Replay Exploration Technique here

### ISSUE THREAD

**Issue by lks9** [lks9/src-tracer#2]
The replay exploration technique is implemented in [lks9/angr](https://github.com/lks9/angr/tree/replayer_exploration_technique). The idea is to move everything in this repo, because it is not likely that this could merge in the angr repo.

* One class for the exploration technique in the `src_trace/` folder
* API methods (perhaps in a different class) to set up everything (all the remaining part which is currently in [replay_trace.py](https://github.com/lks9/src-tracer/blob/replayer_exploration_technique/replay_trace.py))

Requires:
#5

**Comment by lks9** [comment]
Hi @ducorduck, can you check out the `new_directory_structure` branch? There is much difference in the naming and in the directory structure but everything is self-contained in the README.md.

You can already start implementing this issue based on the `new_directory_structure` branch now. Ask me if you have questions.

Thanks!

**Comment by ducorduck** [comment]
Great! Would you merge it to main?

**Comment by lks9** [comment]
done

**Comment by lks9** [comment]
> in order for our the exploration technique to work, the ExplorationTechnique class in angr has to import our exploration technique

Have you checked that this is really necessary? As far as I understand it, you need it there because you call it this way in the script:

```python
    replayer = angr.exploration_techniques.replayer.Replayer(trace_str, functions)
    simgr.use_technique(replayer)
```

I thought you could simply move the exploration technique into a new folder and use it like this:

```python
    replayer = src_tracer.replayer.Replayer(trace_str, functions)
    simgr.use_technique(replayer)
```

>  This changes will likely not be merged into upstream angr therefore if user want to use our exploration technique they would have to modify angr too. Is it a problem?

It makes maintanance harde, because angr frequently gets new versions and we would a fork in sync with the current version of angr. Every change we made in angr so far got either merged or into a separate repo.

**Comment by ducorduck** [comment]
> > in order for our the exploration technique to work, the ExplorationTechnique class in angr has to import our exploration technique
> 
> Have you checked that this is really necessary? As far as I understand it, you need it there because you call it this way in the script:
> 
> ```python
>     replayer = angr.exploration_techniques.replayer.Replayer(trace_str, functions)
>     simgr.use_technique(replayer)
> ```
> 
> I thought you could simply move the exploration technique into a new folder and use it like this:
> 
> ```python
>     replayer = src_tracer.replayer.Replayer(trace_str, functions)
>     simgr.use_technique(replayer)
> ```
> 
you're right i thought the import is needed for something else but it also works like what you've written.

**Comment by lks9** [comment]
I finally found out how we can do something about look ahead into future elements of an iterator: Use [peekable from more-itertools](https://more-itertools.readthedocs.io/en/stable/api.html#more_itertools.peekable). So instead of `next(iter)` just can also use `iter.peek()` to get the next element.
